<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><link rel=icon href=../../favicon.png><meta name=generator content="mkdocs-1.5.3, mkdocs-material-9.5.18"><title>Gene Network Analysis - ReCoDE Home</title><link rel=stylesheet href=../../assets/stylesheets/main.66ac8b77.min.css><link rel=stylesheet href=../../assets/stylesheets/palette.06af60db.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../../stylesheets/extra.css><script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><script id=__analytics>function __md_analytics(){function n(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],n("js",new Date),n("config","G-3S5ER2RYHL"),document.addEventListener("DOMContentLoaded",function(){document.forms.search&&document.forms.search.query.addEventListener("blur",function(){this.value&&n("event","search",{search_term:this.value})}),document$.subscribe(function(){var a=document.forms.feedback;if(void 0!==a)for(var e of a.querySelectorAll("[type=submit]"))e.addEventListener("click",function(e){e.preventDefault();var t=document.location.pathname,e=this.getAttribute("data-md-value");n("event","feedback",{page:t,data:e}),a.firstElementChild.disabled=!0;e=a.querySelector(".md-feedback__note [data-md-value='"+e+"']");e&&(e.hidden=!1)}),a.hidden=!1}),location$.subscribe(function(e){n("config","G-3S5ER2RYHL",{page_path:e.pathname})})});var e=document.createElement("script");e.async=!0,e.src="https://www.googletagmanager.com/gtag/js?id=G-3S5ER2RYHL",document.getElementById("__analytics").insertAdjacentElement("afterEnd",e)}</script><script>var consent;"undefined"==typeof __md_analytics||(consent=__md_get("__consent"))&&consent.analytics&&__md_analytics()</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#single-cell-gene-network-analysis class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--shadow md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../.. title="ReCoDE Home" class="md-header__button md-logo" aria-label="ReCoDE Home" data-md-component=logo> <img src=../../favicon.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> ReCoDE Home </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Gene Network Analysis </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"/></svg> </label> </form> <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../.. class=md-tabs__link> Exemplars </a> </li> <li class=md-tabs__item> <a href=../../about/ class=md-tabs__link> About </a> </li> <li class=md-tabs__item> <a href=../../getting-started/ class=md-tabs__link> Getting started </a> </li> <li class=md-tabs__item> <a href=../../tags/ class=md-tabs__link> Topics </a> </li> <li class=md-tabs__item> <a href=../../coming-soon/ class=md-tabs__link> Coming Soon </a> </li> <li class=md-tabs__item> <a href=../../people/ class=md-tabs__link> People </a> </li> <li class=md-tabs__item> <a href=../../contribute/ class=md-tabs__link> Contributing </a> </li> <li class=md-tabs__item> <a href=../../developing/ class=md-tabs__link> Developing </a> </li> <li class=md-tabs__item> <a href=../../contact/ class=md-tabs__link> Contact </a> </li> <li class=md-tabs__item> <a href=../../licence/ class=md-tabs__link> Licence </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../.. title="ReCoDE Home" class="md-nav__button md-logo" aria-label="ReCoDE Home" data-md-component=logo> <img src=../../favicon.png alt=logo> </a> ReCoDE Home </label> <ul class=md-nav__list data-md-scrollfix> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_1> <div class="md-nav__link md-nav__container"> <a href=../.. class="md-nav__link "> <span class=md-ellipsis> Exemplars </span> </a> <label class="md-nav__link " for=__nav_1 id=__nav_1_label tabindex=0> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_1_label aria-expanded=false> <label class=md-nav__title for=__nav_1> <span class="md-nav__icon md-icon"></span> Exemplars </label> <ul class=md-nav__list data-md-scrollfix> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_1_2> <div class="md-nav__link md-nav__container"> <a href=../../filter/ class="md-nav__link "> <span class=md-ellipsis> All exemplars </span> </a> <label class="md-nav__link " for=__nav_1_2 id=__nav_1_2_label tabindex=0> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_1_2_label aria-expanded=false> <label class=md-nav__title for=__nav_1_2> <span class="md-nav__icon md-icon"></span> All exemplars </label> <ul class=md-nav__list data-md-scrollfix> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_1_2_2> <label class=md-nav__link for=__nav_1_2_2 id=__nav_1_2_2_label tabindex=0> <span class=md-ellipsis> Language </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_1_2_2_label aria-expanded=false> <label class=md-nav__title for=__nav_1_2_2> <span class="md-nav__icon md-icon"></span> Language </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../filter/language/bash/ class=md-nav__link> <span class=md-ellipsis> bash </span> </a> </li> <li class=md-nav__item> <a href=../../filter/language/c%2B%2B/ class=md-nav__link> <span class=md-ellipsis> C++ </span> </a> </li> <li class=md-nav__item> <a href=../../filter/language/c/ class=md-nav__link> <span class=md-ellipsis> C </span> </a> </li> <li class=md-nav__item> <a href=../../filter/language/fortran/ class=md-nav__link> <span class=md-ellipsis> Fortran </span> </a> </li> <li class=md-nav__item> <a href=../../filter/language/julia/ class=md-nav__link> <span class=md-ellipsis> Julia </span> </a> </li> <li class=md-nav__item> <a href=../../filter/language/python/ class=md-nav__link> <span class=md-ellipsis> Python </span> </a> </li> <li class=md-nav__item> <a href=../../filter/language/r/ class=md-nav__link> <span class=md-ellipsis> R </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_1_2_3> <label class=md-nav__link for=__nav_1_2_3 id=__nav_1_2_3_label tabindex=0> <span class=md-ellipsis> Discipline </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_1_2_3_label aria-expanded=false> <label class=md-nav__title for=__nav_1_2_3> <span class="md-nav__icon md-icon"></span> Discipline </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../filter/discipline/astrophysics/ class=md-nav__link> <span class=md-ellipsis> Astrophysics </span> </a> </li> <li class=md-nav__item> <a href=../../filter/discipline/behavioral_science/ class=md-nav__link> <span class=md-ellipsis> Behavioral Science </span> </a> </li> <li class=md-nav__item> <a href=../../filter/discipline/bioinformatics/ class=md-nav__link> <span class=md-ellipsis> Bioinformatics </span> </a> </li> <li class=md-nav__item> <a href=../../filter/discipline/chemical_engineering/ class=md-nav__link> <span class=md-ellipsis> Chemical Engineering </span> </a> </li> <li class=md-nav__item> <a href=../../filter/discipline/computer_science/ class=md-nav__link> <span class=md-ellipsis> Computer Science </span> </a> </li> <li class=md-nav__item> <a href=../../filter/discipline/electrical_engineering/ class=md-nav__link> <span class=md-ellipsis> Electrical Engineering </span> </a> </li> <li class=md-nav__item> <a href=../../filter/discipline/environmental_science/ class=md-nav__link> <span class=md-ellipsis> Environmental Science </span> </a> </li> <li class=md-nav__item> <a href=../../filter/discipline/epidemiology/ class=md-nav__link> <span class=md-ellipsis> Epidemiology </span> </a> </li> <li class=md-nav__item> <a href=../../filter/discipline/finance/ class=md-nav__link> <span class=md-ellipsis> Finance </span> </a> </li> <li class=md-nav__item> <a href=../../filter/discipline/fluid_dynamics/ class=md-nav__link> <span class=md-ellipsis> Fluid Dynamics </span> </a> </li> <li class=md-nav__item> <a href=../../filter/discipline/genetics/ class=md-nav__link> <span class=md-ellipsis> Genetics </span> </a> </li> <li class=md-nav__item> <a href=../../filter/discipline/machine_learning/ class=md-nav__link> <span class=md-ellipsis> Machine Learning </span> </a> </li> <li class=md-nav__item> <a href=../../filter/discipline/mathematics/ class=md-nav__link> <span class=md-ellipsis> Mathematics </span> </a> </li> <li class=md-nav__item> <a href=../../filter/discipline/medicine/ class=md-nav__link> <span class=md-ellipsis> Medicine </span> </a> </li> <li class=md-nav__item> <a href=../../filter/discipline/nuclear_engineering/ class=md-nav__link> <span class=md-ellipsis> Nuclear Engineering </span> </a> </li> <li class=md-nav__item> <a href=../../filter/discipline/patent_analysis/ class=md-nav__link> <span class=md-ellipsis> Patent Analysis </span> </a> </li> <li class=md-nav__item> <a href=../../filter/discipline/physics/ class=md-nav__link> <span class=md-ellipsis> Physics </span> </a> </li> <li class=md-nav__item> <a href=../../filter/discipline/physiology/ class=md-nav__link> <span class=md-ellipsis> Physiology </span> </a> </li> <li class=md-nav__item> <a href=../../filter/discipline/population_dynamics/ class=md-nav__link> <span class=md-ellipsis> Population Dynamics </span> </a> </li> <li class=md-nav__item> <a href=../../filter/discipline/quantum_mechanics/ class=md-nav__link> <span class=md-ellipsis> Quantum Mechanics </span> </a> </li> <li class=md-nav__item> <a href=../../filter/discipline/robotics/ class=md-nav__link> <span class=md-ellipsis> Robotics </span> </a> </li> <li class=md-nav__item> <a href=../../filter/discipline/software_engineering/ class=md-nav__link> <span class=md-ellipsis> Software Engineering </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_1_2_4> <label class=md-nav__link for=__nav_1_2_4 id=__nav_1_2_4_label tabindex=0> <span class=md-ellipsis> Method </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_1_2_4_label aria-expanded=false> <label class=md-nav__title for=__nav_1_2_4> <span class="md-nav__icon md-icon"></span> Method </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../filter/method/bayesian_inference/ class=md-nav__link> <span class=md-ellipsis> Bayesian Inference </span> </a> </li> <li class=md-nav__item> <a href=../../filter/method/computer_vision/ class=md-nav__link> <span class=md-ellipsis> Computer Vision </span> </a> </li> <li class=md-nav__item> <a href=../../filter/method/data_analysis/ class=md-nav__link> <span class=md-ellipsis> Data Analysis </span> </a> </li> <li class=md-nav__item> <a href=../../filter/method/deep_learning/ class=md-nav__link> <span class=md-ellipsis> Deep Learning </span> </a> </li> <li class=md-nav__item> <a href=../../filter/method/embedded_systems/ class=md-nav__link> <span class=md-ellipsis> Embedded Systems </span> </a> </li> <li class=md-nav__item> <a href=../../filter/method/gui_development/ class=md-nav__link> <span class=md-ellipsis> GUI Development </span> </a> </li> <li class=md-nav__item> <a href=../../filter/method/machine_learning/ class=md-nav__link> <span class=md-ellipsis> Machine Learning </span> </a> </li> <li class=md-nav__item> <a href=../../filter/method/natural_language_processing/ class=md-nav__link> <span class=md-ellipsis> Natural Language Processing </span> </a> </li> <li class=md-nav__item> <a href=../../filter/method/numerical_computing/ class=md-nav__link> <span class=md-ellipsis> Numerical Computing </span> </a> </li> <li class=md-nav__item> <a href=../../filter/method/numerical_methods/ class=md-nav__link> <span class=md-ellipsis> Numerical Methods </span> </a> </li> <li class=md-nav__item> <a href=../../filter/method/sequence_analysis/ class=md-nav__link> <span class=md-ellipsis> Sequence Analysis </span> </a> </li> <li class=md-nav__item> <a href=../../filter/method/statistical_analysis/ class=md-nav__link> <span class=md-ellipsis> Statistical Analysis </span> </a> </li> <li class=md-nav__item> <a href=../../filter/method/survival_analysis/ class=md-nav__link> <span class=md-ellipsis> Survival Analysis </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../../about/ class=md-nav__link> <span class=md-ellipsis> About </span> </a> </li> <li class=md-nav__item> <a href=../../getting-started/ class=md-nav__link> <span class=md-ellipsis> Getting started </span> </a> </li> <li class=md-nav__item> <a href=../../tags/ class=md-nav__link> <span class=md-ellipsis> Topics </span> </a> </li> <li class=md-nav__item> <a href=../../coming-soon/ class=md-nav__link> <span class=md-ellipsis> Coming Soon </span> </a> </li> <li class=md-nav__item> <a href=../../people/ class=md-nav__link> <span class=md-ellipsis> People </span> </a> </li> <li class=md-nav__item> <a href=../../contribute/ class=md-nav__link> <span class=md-ellipsis> Contributing </span> </a> </li> <li class=md-nav__item> <a href=../../developing/ class=md-nav__link> <span class=md-ellipsis> Developing </span> </a> </li> <li class=md-nav__item> <a href=../../contact/ class=md-nav__link> <span class=md-ellipsis> Contact </span> </a> </li> <li class=md-nav__item> <a href=../../licence/ class=md-nav__link> <span class=md-ellipsis> Licence </span> </a> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#description class=md-nav__link> <span class=md-ellipsis> Description </span> </a> </li> <li class=md-nav__item> <a href=#overall-learning-outcomes class=md-nav__link> <span class=md-ellipsis> Overall Learning Outcomes </span> </a> </li> <li class=md-nav__item> <a href=#requirements class=md-nav__link> <span class=md-ellipsis> Requirements </span> </a> </li> <li class=md-nav__item> <a href=#getting-started class=md-nav__link> <span class=md-ellipsis> Getting Started </span> </a> <nav class=md-nav aria-label="Getting Started"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#cloning-the-repository class=md-nav__link> <span class=md-ellipsis> Cloning the repository </span> </a> </li> <li class=md-nav__item> <a href=#setting-up-your-environment class=md-nav__link> <span class=md-ellipsis> Setting up your environment </span> </a> </li> <li class=md-nav__item> <a href=#system class=md-nav__link> <span class=md-ellipsis> System </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#development class=md-nav__link> <span class=md-ellipsis> Development </span> </a> <nav class=md-nav aria-label=Development> <ul class=md-nav__list> <li class=md-nav__item> <a href=#setting-up-pre-commit class=md-nav__link> <span class=md-ellipsis> Setting up pre-commit </span> </a> </li> <li class=md-nav__item> <a href=#managing-dependencies-with-pip-tools class=md-nav__link> <span class=md-ellipsis> Managing Dependencies with pip-tools </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#project-structure class=md-nav__link> <span class=md-ellipsis> Project Structure </span> </a> </li> <li class=md-nav__item> <a href=#license class=md-nav__link> <span class=md-ellipsis> License </span> </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <nav class=md-tags> <a href=../../tags/#single-cell-rna-sequencing class=md-tag>Single-Cell RNA Sequencing</a> <a href=../../tags/#methods-for-single-cell-gene-network-analysis class=md-tag>Methods for Single-Cell Gene Network Analysis</a> <a href=../../tags/#gene-set-enrichment-analysis class=md-tag>Gene Set Enrichment Analysis</a> <a href=../../tags/#visualisation class=md-tag>Visualisation</a> <a href=../../tags/#natural-language-processing class=md-tag>Natural language processing</a> <a href=../../tags/#data-analysis class=md-tag>Data analysis</a> </nav> <div id=overlay> <div id=overlay-text> Redirecting to exemplar docs... </div> </div> <p><meta http-equiv=refresh content="0; url=https://imperialcollegelondon.github.io/ReCoDE-Gene-Network-Analysis"></p> <!-- Your Project title, make it sound catchy! --> <h1 id=single-cell-gene-network-analysis>Single-Cell Gene Network Analysis<a class=headerlink href=#single-cell-gene-network-analysis title="Permanent link">&para;</a></h1> <!-- Provide a short description to your project --> <h2 id=description>Description<a class=headerlink href=#description title="Permanent link">&para;</a></h2> <p>The aim of this exemplar is to understand and demonstrate how to carry out single-cell gene network analysis on real data. Specifically, we will model networks in a small, publicly available Human Cancer dataset as well as explore them further through downstream analysis. This will consist of characterising the networks through gene set enrichment analysis and understanding the principles behind this. The modelling itself will encompass learning how to use both correlation-based approaches as well as deep learning techniques, such as large language models to construct networks. Together, this will provide an undertanding as well as practical experience with an end-to-end pipeline using single-cell data for gene network analysis.</p> <!-- What should the students going through your exemplar learn --> <h2 id=overall-learning-outcomes>Overall Learning Outcomes<a class=headerlink href=#overall-learning-outcomes title="Permanent link">&para;</a></h2> <p>By the end of the series of chapters and notebooks you should be able to:</p> <ol> <li>Understand the Foundations of Single-Cell RNA-Sequencing (scRNA-seq):</li> <li>Comprehend the steps involved in scRNA-seq data processing, including generating count matrices and performing quality control.</li> <li>Explain how scRNA-seq differs from bulk RNA-seq in capturing cellular heterogeneity.</li> <li>Explore publicly available snRNA-seq data and process it for downstream analysis using Python.</li> <li>Apply Techniques in Gene Network Analysis:</li> <li>Understand the concepts and methods involved in constructing gene networks from scRNA-seq data.</li> <li>Identify relationships between genes using co-expression based approaches.</li> <li>Visualise gene networks using tools such as NetworkX.</li> <li>Carry Out Gene Set Enrichment Analysis:</li> <li>Explain the significance of gene set enrichment and its role in interpreting gene networks.</li> <li>Perform pathway-based enrichment analyses using curated gene sets from resources like KEGG and Gene Ontology (GO).</li> <li>Apply statistical tests such as the Fisher Exact Test to assess the overrepresentation of gene sets.</li> <li>Implement multiple hypothesis correction methods like the false discovery rate (FDR) to ensure robust enrichment results.</li> <li>Utilise Large Language Models for Gene Network Predictions:</li> <li>Explain the basic principles of NLP, including tokenization, stemming, and dependency parsing.</li> <li>Recognize the capabilities of LLMs in analyzing complex biological datasets, even when data is limited.</li> <li>Apply pretrained models like scGPT to predict gene interactions and network dynamics in scRNA-seq data.</li> <li>Understand the role of transformers and attention mechanisms in enhancing biological predictions through context-aware learning.</li> </ol> <!-- How long should they spend reading and practising using your Code.
Provide your best estimate --> <table> <thead> <tr> <th>Task</th> <th>Time</th> </tr> </thead> <tbody> <tr> <td>Reading</td> <td>3 hours</td> </tr> <tr> <td>Practising</td> <td>3 hours</td> </tr> </tbody> </table> <h2 id=requirements>Requirements<a class=headerlink href=#requirements title="Permanent link">&para;</a></h2> <!--
--> <ul> <li>Familiarity with Python</li> <li>Basic knowledge on NLP and LLMs</li> <li>Basic knowledge on single-cell RNA-sequencing</li> </ul> <h2 id=getting-started>Getting Started<a class=headerlink href=#getting-started title="Permanent link">&para;</a></h2> <p>Take a look at the table of contents below and follow the chapters from 1 to 4. These are also aligned with the notebooks and should give you background information as well as direction and motivation behind the notebooks. If you are ever unsure or don't understand how to do something on the workbooks, there are also fully worked through solutions that you can use.</p> <h3 id=cloning-the-repository>Cloning the repository<a class=headerlink href=#cloning-the-repository title="Permanent link">&para;</a></h3> <p>If you don't have it already, install git to your machine, see <a href=https://git-scm.com/book/en/v2/Getting-Started-Installing-Git>here</a> for details on all OS's.</p> <p>Once installed, run the following to clone the repository and change the working directory to the clones repository:</p> <div class=highlight><pre><span></span><code>git<span class=w> </span>clone<span class=w> </span>https://github.com/ImperialCollegeLondon/ReCoDE-Gene-Network-Analysis.git
<span class=nb>cd</span><span class=w> </span>ReCoDE-Gene-Network-Analysis
</code></pre></div> <h3 id=setting-up-your-environment>Setting up your environment<a class=headerlink href=#setting-up-your-environment title="Permanent link">&para;</a></h3> <p>Before installing the project dependencies, it's recommended to set up a virtual environment. This ensures that the dependencies for this project do not interfere with those of other Python projects you may be working on. Follow these steps to create and activate a virtual environment:</p> <ol> <li><strong>Create a Virtual Environment</strong></li> </ol> <p>If you haven't already, navigate to the root directory of the project and run the following command to create a virtual environment. This command creates a new directory <code>.venv</code> which will contain all the necessary executables to use the packages that a Python project would need.</p> <div class=highlight><pre><span></span><code>python3<span class=w> </span>-m<span class=w> </span>venv<span class=w> </span>.venv
</code></pre></div> <ol> <li><strong>Activate the Virtual Environment</strong></li> </ol> <p>After creating the virtual environment, you need to activate it. Activation of the virtual environment will change your shell’s prompt to show what virtual environment you’re using, and modify the environment so that running python will get you that particular version and installation of Python.</p> <ul> <li><strong>On Windows</strong></li> </ul> <div class=highlight><pre><span></span><code>.<span class=se>\v</span>env<span class=se>\S</span>cripts<span class=se>\a</span>ctivate
</code></pre></div> <ul> <li><strong>On Unix or MacOS</strong></li> </ul> <div class=highlight><pre><span></span><code><span class=nb>source</span><span class=w> </span>venv_recode_GeneNetworkAnalysis/bin/activate
</code></pre></div> <p>This command needs to be run every time you start a new terminal session and want to work on this project.</p> <ol> <li><strong>Install Dependencies</strong></li> </ol> <p>With the virtual environment activated, install the project dependencies by running:</p> <div class=highlight><pre><span></span><code>pip<span class=w> </span>install<span class=w> </span>-r<span class=w> </span>requirements.txt
</code></pre></div> <p>Remember to deactivate your virtual environment when you're done working on the project by running deactivate.</p> <!-- An overview of the files and folder in the exemplar.
Not all files and directories need to be listed, just the important
sections of your project, like the learning material, the code, the tests, etc.

A good starting point is using the command `tree` in a terminal(Unix),
copying its output and then removing the unimportant parts.

You can use ellipsis (...) to suggest that there are more files or folders
in a tree node.

--> <h3 id=system>System<a class=headerlink href=#system title="Permanent link">&para;</a></h3> <!-- Instructions on how the student should start going through the exemplar.

Each jupyter notebook workbook is paired with a chapter of information ordered from 1 to 4. The notebooks can be accessed within the notebooks folder and the chapters can be accessed within the docs folder. Upon completion, you can go through the solutions notebooks which are also stored in the docs folder. 

--> <h2 id=development>Development<a class=headerlink href=#development title="Permanent link">&para;</a></h2> <p>This project uses <code>pre-commit</code> and <code>pip-tools</code> for managing pre-commit hooks and dependencies respectively. Follow these steps to set up your development environment:</p> <h3 id=setting-up-pre-commit>Setting up pre-commit<a class=headerlink href=#setting-up-pre-commit title="Permanent link">&para;</a></h3> <ol> <li>Install pre-commit on your system. If you haven't installed it yet, you can do so by running:</li> </ol> <div class=highlight><pre><span></span><code>pip<span class=w> </span>install<span class=w> </span>pre-commit
</code></pre></div> <ol> <li>Once installed, you can set up the pre-commit hooks by running the following command in the root directory of this project:</li> </ol> <div class=highlight><pre><span></span><code>pre-commit<span class=w> </span>install
</code></pre></div> <p>This will install all the pre-commit hooks defined in .pre-commit-config.yaml into your local repository. These hooks will automatically run on every commit to ensure code quality and consistency.</p> <h3 id=managing-dependencies-with-pip-tools>Managing Dependencies with pip-tools<a class=headerlink href=#managing-dependencies-with-pip-tools title="Permanent link">&para;</a></h3> <p>Dependencies are managed using the <a href=https://github.com/jazzband/pip-tools>pip-tools</a> tool chain.</p> <p>Unpinned dependencies are specified in <code>pyproject.toml</code>. Pinned versions are then produced with:</p> <div class=highlight><pre><span></span><code>pip-compile<span class=w> </span>pyproject.toml
</code></pre></div> <p>To add/remove packages edit <code>pyproject.toml</code> and run the above command. To upgrade all existing dependencies run:</p> <div class=highlight><pre><span></span><code>pip-compile<span class=w> </span>--upgrade<span class=w> </span>pyproject.toml
</code></pre></div> <p>Dependencies for developers are listed separately as optional, with the pinned versions being saved to <code>dev-requirements.txt</code> instead.</p> <p><code>pip-tools</code> can also manage these dependencies by adding extra arguments, e.g.:</p> <div class=highlight><pre><span></span><code>pip-compile<span class=w> </span>-o<span class=w> </span>dev-requirements.txt<span class=w> </span>--extra<span class=o>=</span>dev<span class=w> </span>pyproject.toml
</code></pre></div> <p>When dependencies are upgraded, both requirements.txt and dev-requirements.txt should be regenerated so that they are in sync.</p> <h2 id=project-structure>Project Structure<a class=headerlink href=#project-structure title="Permanent link">&para;</a></h2> <div class=highlight><pre><span></span><code>.
├── .github/workflows
├── data
|   ├── data
|       ├── Bcell_datExpr_pseudobulk.csv
│       └── Bcell_filtered_compressed.h5ad.gz
|   ├── metadata
|       ├── Bcell_metadata.csv
│       └── Bcell_metadata_pseudobulk.csv
│   └── other
|       ├── h.all.v2023.2.Hs.symbols.gmt
│       └── separated_communities.pkl
├── docs
|   ├── .icons/logos
|   ├── assets
|   ├── Chapter 1: Background on Single-Cell RNA-Sequencing
|   ├── Chapter 2: Gene Network Analysis
|   ├── Chapter 3: Gene Network Downstream Analysis
│   └── Chapter 4: Understanding NLP and LLMs
├── notebooks
|   ├── solutions
|       ├── 1.Correlation_Network_Analysis_and_Visualisation_Solution.ipynb
|       ├── 2.Network_Analysis_Solution.ipynb
|       ├── 3.Fisher_Exact_Test_Solution.ipynb
│       └── 4.SCGPT_Solution.ipynb
|   ├── workbooks
|       ├── 1.Correlation_Network_Analysis_and_Visualisation_Workbook.ipynb
|       ├── 2.Network_Analysis_Workbook.ipynb
|       ├── 3.Fisher_Exact_Test_Workbook.ipynb
│       └── 4.SCGPT_Workbook.ipynb
|   ├── .placeholder
|   ├── process_data.ipynb
├── test
|   ├── test_placeholder.py
├── .gitignore
├── .pre-commit-config.yaml
├── LICENSE.md # Outlines what legal rights you have to use this software.
├── README.md  # You are here!
├── dev-requirements.txt # Development requirements
├── mkdocs.yml # Tells readthedocs.com how to build the documentation.
├── pyproject.toml # Machine readable information about the package.
├── requirements.txt # Requirements for the notebooks.
└── test
</code></pre></div> <h2 id=license>License<a class=headerlink href=#license title="Permanent link">&para;</a></h2> <p>This project is licensed under the <a href=LICENSE.md>BSD-3-Clause license</a></p> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </button> </main> <footer class=md-footer> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> &copy; 2022 - 2025, Imperial College, London. All rights reserved. – <a href=#__consent>Change cookie settings</a> </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/ImperialCollegeLondon/ReCoDE-home target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <div class=md-consent data-md-component=consent id=__consent hidden> <div class=md-consent__overlay></div> <aside class=md-consent__inner> <form class="md-consent__form md-grid md-typeset" name=consent> <h4>Cookie consent</h4> <p>We use cookies to recognize your repeated visits and preferences, as well as to measure the effectiveness of our documentation and whether users find what they're searching for. With your consent, you're helping us to make our documentation better.</p> <input class=md-toggle type=checkbox id=__settings> <div class=md-consent__settings> <ul class=task-list> <li class=task-list-item> <label class=task-list-control> <input type=checkbox name=analytics checked> <span class=task-list-indicator></span> Google Analytics </label> </li> </ul> </div> <div class=md-consent__controls> <button class="md-button md-button--primary">Accept</button> <label class=md-button for=__settings>Manage settings</label> </div> </form> </aside> </div> <script>var consent=__md_get("__consent");if(consent)for(var input of document.forms.consent.elements)input.name&&(input.checked=consent[input.name]||!1);else"file:"!==location.protocol&&setTimeout(function(){document.querySelector("[data-md-component=consent]").hidden=!1},250);var action,form=document.forms.consent;for(action of["submit","reset"])form.addEventListener(action,function(e){if(e.preventDefault(),"reset"===e.type)for(var n of document.forms.consent.elements)n.name&&(n.checked=!1);__md_set("__consent",Object.fromEntries(Array.from(new FormData(form).keys()).map(function(e){return[e,!0]}))),location.hash="",location.reload()})</script> <script id=__config type=application/json>{"base": "../..", "features": ["navigation.instant", "navigation.top", "navigation.indexes", "navigation.tracking", "navigation.tabs", "navigation.tabs.sticky", "navigation.sections", "search.highlight", "search.share", "search.suggest", "toc.follow"], "search": "../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script> <script src=../../assets/javascripts/bundle.3220b9d7.min.js></script> </body> </html>